File: solutions/elixir/anagram/3/lib/anagram.ex
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t(), [String.t()]) :: [String.t()]
  def match(base, candidates) do
    candidates
    |> Enum.filter(&anagram?(String.downcase(base), String.downcase(&1)))
  end

  defp anagram?(base, candidate) when base == candidate, do: false

  defp anagram?(base, candidate) do
    normalized_base = normalize(base)
    normalized_candidate = normalize(candidate)
    normalized_base == normalized_candidate
  end

  defp normalize(word) do
    word
    |> String.graphemes()
    |> Enum.sort()
  end
end

File: solutions/ocaml/anagram/3/anagram.ml
let normalize x =
  x |> String.lowercase_ascii |> String.to_seq |> List.of_seq |> List.sort Char.compare
;;

let anagrams target =
  let target_lower = String.lowercase_ascii target in
  let normalized_target = normalize target in
  let is_anagram c =
    let cand_lower = String.lowercase_ascii c in
    cand_lower <> target_lower && normalized_target = normalize c
  in
  List.filter is_anagram
;;
